src/Hadolint/Rules.hs

```diff
@@ -89,12 +89,19 @@ type Rule = ParsedFile -> [RuleCheck]
 mapInstructions :: CheckerWithState state -> state -> Rule
 mapInstructions f initialState dockerfile =
   let (_, results) = mapAccumL applyRule initialState dockerfile
-   in concat results
+      (_, nestedResults) = mapAccumL nestedApplyRule initialState dockerfile
+   in concat $ results ++ [r | r <- nestedResults, r `notElem` results]
   where
-    applyRule state (InstructionPos (OnBuild i) source linenumber) =
```

Comment:
I believe this could be written without the second pass and the merging of the rules by calling `applyWithState` twice on the `OnBuild` case like this:

```
    applyRule state (InstructionPos onbuild@(OnBuild i) source linenumber) =
      -- All rules applying to instructions also apply to ONBUILD,
      -- so we unwrap the OnBuild constructor and check directly the inner
      -- instruction
      let (innerState, innerResults) = applyWithState state source linenumber i
          (finalState, outerResults) = applyWithState innerState source linenumber onbuild
       in (finalState, innerResults <> outerResults)
```

Reply:
This is great! This is exactly what I had in mind. Thank you for showing me :-)